src/amoc_api_scenarios_handler.erl

```diff
@@ -10,112 +10,128 @@
+malformed_request(Req, State) ->
+    {Method, Req2} = cowboy_req:method(Req),
+    malformed_request(Method, Req2, State).
+
+malformed_request(<<"POST">>, Req, State) ->
+    {ok, Body, Req2} = cowboy_req:body(Req),
+    case jsx:is_json(Body) of
+        false ->
+            {true, Req2, State};
+        true ->
+            JSON = jsx:decode(Body),
+            ContainModuleName = proplists:is_defined(<<"scenario">>, JSON),
+            ContainModuleSource = proplists:is_defined(<<"module_source">>, JSON),
+
+            case ContainModuleName and ContainModuleSource of
+                true ->
+                    ModuleName = proplists:get_value(<<"scenario">>, JSON),
+                    ModuleSource = proplists:get_value(<<"module_source">>, JSON),
+                    State2 = State#state{
+                               module_name = ModuleName,
+                               module_source = ModuleSource
+                               },
+                    {false, Req2, State2};
+                false ->
+                    {true, Req2, State}
+            end
     end;
 
-process_json(_Term, #state{action = stop}) ->
-    {200, jsx:encode("ok")};
-
-process_json(Term, #state{action = load}) ->
-    ScenarioB = proplists:get_value(<<"scenario">>, Term),
-    ModuleB = proplists:get_value(<<"module_source">>, Term),
-    Scenario = erlang:binary_to_atom(ScenarioB, utf8),
-    ScenarioPath = "scenarios/" ++ erlang:atom_to_list(Scenario) ++ ".erl",
-    ok = file:write_file(ScenarioPath, ModuleB, [write]),
-    case compile:file(ScenarioPath, [{outdir, "ebin"}]) of
-        {ok, Scenario} ->
-            code:purge(Scenario),
-            {200, jsx:encode([{ScenarioB, <<"loaded">>}])};
-        error ->
-            {500, jsx:encode([{<<"error">>, <<"compilation_error">>}])}
-    end;
+malformed_request(_, Req, State) ->
+    {false, Req, State}.
 
-process_json(_Term, #state{action = ping_nodes}) ->
-    Nodes = amoc_dist:ping_nodes(),
-    {200, jsx:encode([{<<"nodes">>, Nodes}])};
 
-process_json(_Term, #state{action = list}) ->
+%% Request processing functions
+
+-spec handle_get(cowboy:req(), state()) -> 
+    {string() | halt, cowboy:req(), state()}.
+handle_get(Req0, State = #state{}) ->
     {ok, Filenames} = file:list_dir("scenarios"),
-    FilenamesB = [ list_to_binary(X) || X <- Filenames ],
-    {200, jsx:encode([{<<"scenarios">>, FilenamesB}])}.
+    Filenames2 =
+        lists:filter(
+          fun(X) -> string:right(X,3) == "erl" end,
+          Filenames),
+
+    Scenarios = 
+    [ erlang:list_to_binary(Y) ||  X <- Filenames2,
+                                   Y <- string:tokens(X, "."),
+                                   Y =/= "erl" ],
+    Reply = jsx:encode([{scenarios,Scenarios}]),
+    {Reply, Req0, State}.
+
+
+-spec handle_post(cowboy:req(), state()) ->
+    {string() | halt | ok, cowboy:req(), state()}.
+handle_post(Req0, State = #state{module_name = ModuleName, module_source = ModuleSource}) ->
```

Comment:
Line 116 is too long: handle_post(Req0, State = #state{module_name = ModuleName, module_source = ModuleSource}) ->.